Refactored the python init function selector

elpaso opened the pull request:
Do not rely on the presence of a dot to load
a module. Loading a module can now be achieved
through the code editor in the dialog.
- added an option to load from environment
- added a file selector to specify the file
- changed tooltips accordingly

This is the new GUI:

![qgis_store_python_code_latest](https://cloud.githubusercontent.com/assets/142164/11364767/b3e10742-92a2-11e5-9d94-14207756dc29.png)

When loading from external file:

![qgis_store_python_code_latest_ext_file](https://cloud.githubusercontent.com/assets/142164/11364799/ed4a0d9e-92a2-11e5-9c68-859c29da5ee9.png)

elpaso committed:
Refactored the python init function selector

Do not rely on the presence of a dot to load
a module. Loading a module can now be achieved
through the code editor in the dialog.

* added an option to load from environment
* added a file selector to specify the file
* changed tooltips accordingly

m-kuhn commented:
Nice.
Does it provide support for existing projects with init functions from external files?
Is the init file path saved relative?

elpaso commented:
@m-kuhn, not really.
You mean temporarily support the "old" "dot" behavior  for backward compatibility? This would also mean to temporary bypass the combo if there's a dot in the function name.
I've also thought about the relative/absolute, but I opted to just clone the behavior currently implemented for the UI specified file.

m-kuhn commented:
Btw, on the screenshots there is the file chooser button still visible when it shouldn't.

elpaso commented:
Oh, yes, forgot to to hide it... thanks

m-kuhn commented:
I think the relative path is a key-feature which is currently supported and it would be bad to remove it.

Concerning the loading, it could be checked on project migration if it's a module path and if yes insert the appropriate file path.

elpaso committed:
Hide file selector when not needed

elpaso commented:
@m-kuhn sorry, I don't follow you. What is "a key-feature" ?  Are you talking about the file selector?
Or do you mean that the "dot" loading mechanism allowed for relative loading of modules?

m-kuhn commented:
I think there are two mandatory parts:

 * It should be possible to open 2.12 projects without having to reconfigure anything (legacy/migration)
 * Having the possibility to put the .py file next to the project file and use it from different network shares/send it in a .zip (relative file paths)

elpaso commented:
Ok, for the first point I think I need some guidance, how it is normally handled? Is there a project import migration routine? 
For the second point: 
1. isn't that the same for the UI file?
2. If yes, why should the .py be handled differently from the .ui file ?
In any case: local loading of python files (old "dot") behavior is now possible with the code provided in the dialo, you can simply "from myform import my_open_func" and call that func, and as long as it's in the path is gets loaded, like in the old "dot" behavior.

Or I completely missed the point and it's just a problem of how the path is stored in the project file and to make it relative?

m-kuhn commented:
1. When loading the xml check if the init source is not yet already set. If not check for a dot. If present extract the file name. Set the init source accordingly.
 2. I agree it should be the same for the .ui file. But if it's not yet implemented there it's not being treated as a regression.

elpaso commented:
Agree for the b/w compatibility fix in qgsvectorlayer.cpp (I hoped there was a better place to implement temporary hacks).

For the relative path I still don't see any regression with the relative path of the python file: this is a new feature.
The old "dot" behavior (load a module, like myform.open ) is now possible with a new more flexible system (the code provided in the dialog): just place 

from my_forms import my_open

in the provided code and "my_open" as init function and you're done.

The use can also choose a python file: this is a new feature: where is the regression?

m-kuhn commented:
I am probably getting it all wrong, sorry.

If I previously had a init function that was `form.open` it would look for the file `form.py` in the same folder as my `project.qgs` and load it. I could copy the two files to another place and it would still work.

Question: If with your patches applied I reopen the project and save it again. And then copy the two files to another location. Will it still work?

elpaso commented:
the old `form.open` is still now possible in the patch by:
* write `open` in the form init,
* choose "Provide code in this dialog" option
* place code in a file `form.py` in the project folder (or in any other python path folder),
* write `from form import open` in the code editor

But I agree that the new feature that allow to specify a file path through the file chooser should take into account the project settings for relative paths (for the UI file too). 

Answering to your latest question, I still need to implement the b/w compatibility that transforms the old "dot" syntax to the new one but when it will be implemented I'd say that the answer is yes:

Opening a 2.12 project with the "dot" notation would transform it splitting on the dot and placing an `from form import open` into the provided code.
Saving the project into 2.12+x format would just do the right thing.

m-kuhn commented:
Nice, that sounds good.

> the old `form.open` is still now possible in the patch by:
* write `open` in the form init,
* choose "Provide code in this dialog" option
* place code in a file `form.py` in the project folder (or in any other python path folder),
* write `from form import open` in the code editor

Or by choosing form.py in the file browser and writing `open` in the init function I guess?

elpaso commented:
Yes, sure, or just write `open` and hope for the best as long as it's provided in the environment, for example by a plugin or `startup.py`

m-kuhn commented:
That should never be a problem I guess
https://docs.python.org/2/library/functions.html#open

@NathanW2 What's your opinion about this?

elpaso commented:
Yeah, I 'd chosen a better example.

m-kuhn commented:
I think it's actually the best example.
Because if this still works the same backwards compatibility should be given for any other case as well.

elpaso commented:
I don't seen any difference form the old `form.open` and the new implementation. In both cases the `open` function defined in the `forms.py` is called. 
Is that what you wanted to make sure?
If there's no `open` function in `forms.py` an attempt to inspect and run a `open(*args)` with len(args)=3 is made, and this is also the same behavior in both implementations.

elpaso committed:
Added b/w compatibility for "dot" module/function

When loading a function init with a dot, this
is transformed to a proper import and the
python init configuration is changed to support
the new python loading system.

Also check for inspect errors.

Still missing: warn the user on configuration and
trapped runtime errors

elpaso commented:
@NathanW2 later on I would also like to some error messages/warnings to the user when something goes wrong (misconfigurations or python errors). Any suggestion?

elpaso committed:
Added error messages with QMessageBox

Errors on:
* init function not found
* init function with wrong parameters count

NathanW2 commented:
I will have a look over this tonight.

Out of scope here: Does anyone have plans to rework this field dialog.
It's really a big mess these days.

On Wed, Nov 25, 2015 at 6:39 PM, Alessandro Pasotti <
notifications@github.com> wrote:

> @NathanW2 <https://github.com/NathanW2> later on I would also like to
> some error messages/warnings to the user when something goes wrong
> (misconfigurations or python errors). Any suggestion?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/2498#issuecomment-159535035>.
>

m-kuhn commented:
@NathanW2 I once did some mockups https://github.com/qgis/QGIS-Enhancement-Proposals/issues/37 but no plans to do something about it right now.

elpaso commented:
@NathanW2 any comment? I'll merge in a few days.

NathanW2 commented:
Looking now.

NathanW2 commented:
This looks ok to me.  

Just to check. The load from environment should allow

`myform.form_open` 

if `myform` is on path as per the current setup?

I'm happy with how this functions although we do need to rework this panel in the dialog at a future date because it's just too much.

If there is no other comments and all tests are passing good for merge.

elpaso commented:
@NathanW2 yes and no: the idea is to avoid the dot (".") in the function name as a selector for the source of python code, now that you can enter the file path and enter the python code directly, this is not necessary anymore, now to load `myform.form_open` you can use the code dialog and enter `from myfrom import form_open` and just write `form_open` as your init function.
But, to maintain backward compatibility and an easy migration I have implemented an import routine that will rewrite `myform.form_open` as indicated above.
So, basically you can safely import older projects and  `myform.form_open` will be transformed to the new configuration format.

See: https://github.com/qgis/QGIS/pull/2498/files#diff-106be592650b55ccae79c83eecd15cebR1795

elpaso committed:
Honour relative path setting
